### DISCUSSION TITLE
Custom fonts plugin

### DISCUSSION THREAD

**Discussion by lexoyo** [GrapesJS/grapesjs#4858] [Show and tell]
> edit: [this plugin is now live here](https://www.npmjs.com/package/@silexlabs/grapesjs-fonts)

Hello

I am developing a "custom fonts" plugin for grapesjs and I need feedback and ideas

I saw discussions and issues like "How can i add custom fonts in grapesjs editor? #4563" 

What seems to work for me is

1. update the "font family" dropdown
    ```
    const styleManager = editor.StyleManager
    const fontProperty = styleManager.getProperty('typography', 'font-family')
    fontProperty.setOptions(fonts)
    styleManager.render()
    ```
1. update the DOM to display the font correctly: add style elements to the editor.Canvas.getDocument()

This is quite easy but here are the things which took me time as I implemented google fonts

* use google fonts api to select fonts and get their name, variants, weights
* build the URL of the fonts to load
* the UI to manage and install fonts

It looks like this:

![Screenshot from 2023-01-20 16-20-56](https://user-images.githubusercontent.com/715377/213734511-7e66175b-cb72-4a61-b215-2af64f5d532c.png)
![Screenshot from 2023-01-20 16-19-41](https://user-images.githubusercontent.com/715377/213734520-adc1072f-ed94-4a01-b1e0-3560a6816083.png)

It has these features

* API to add / remove fonts from the site (from goole font name) 
* Updates the DOM and the "font family" dropdown
* Save the fonts with the site data
* Load the fonts when site data is loaded (add to the DOM on load)
* UI to manage fonts
* Integration with google API

What do you think? Is this useful? Is there too much or too few features? Is google fonts a good idea?

**Comment by artf** [comment]
Looks awesome @lexoyo 👏

**Comment by lexoyo** [comment]
This plugin is now [live here](https://www.npmjs.com/package/@silexlabs/grapesjs-fonts)

**Comment by kyr0** [comment]
Thank you @lexoyo !

**Comment by kyr0** [comment]
@lexoyo why not adding it to `README.md` here just like the other PR https://github.com/artf/grapesjs/pull/4746 merged a few days ago? I think it would make the plugin a bit more popular and it deserves it :)

**Reply by lexoyo** [reply]
Hey thank you @kyr0 🎉

I hope it does, then someone may have an idea of how to use Google fonts v3 with support of variable fonts

I will add it like the other plugin, good idea

**Reply by lexoyo** [reply]
Done
https://github.com/artf/grapesjs/pull/4877
:fireworks:

**Comment by voleep** [comment]
Woww, so appreciate. Thank you!

**Comment by RakulAgn** [comment]
hey @lexoyo fontProperty.setOptions(fonts). is not exists in the getProperty it shows type error
```
Property 'setOptions' does not exist on type 'Property<PropertyProps>'.
```

this is want i am asking about

**Reply by lexoyo** [reply]
i answered you here already
https://github.com/silexlabs/grapesjs-fonts/issues/16#issuecomment-2296035066
